staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,1011 @@
+
+	// NodeSelector defines which nodes have access to the resources in the pool.
+	// If it is specified, but empty (has no terms), all nodes have access.
+	//
+	// Exactly one of NodeName and NodeSelector must be set.
+	//
+	// +optional
+	NodeSelector *v1.NodeSelector `json:"nodeSelector,omitempty" protobuf:"bytes,4,opt,name=nodeSelector"`
+
+	// Devices lists some or all of the devices in this pool.
+	//
+	// Must not have more than 128 entries.
+	//
+	// +optional
+	// +listType=atomic
+	Devices []Device `json:"devices" protobuf:"bytes,5,name=devices"`
+}
+
+// ResourcePool describes the pool that ResourceSlices belong to.
+type ResourcePool struct {
+	// Name is used to identify the pool. For node-local devices, this
+	// is often the node name, but this is not required.
+	//
+	// It must not be longer than 253 characters and must consist of one or more DNS sub-domains
+	// separated by slashes.
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// Generation must get increased in all ResourceSlices of a pool whenever resource
+	// definitions change. A consumer must only use definitions from ResourceSlices
+	// with the highest generation number and ignore all others.
+	//
+	// This mechanism enables consumers to detect pools that are in an incomplete state.
+	// It is not sufficient to determine whether the pool is up-to-date. Because that
+	// race is unavoidable, drivers may start publishing a pool with zero as initial
+	// generation number when there are no slices for that pool with a higher number.
+	//
+	// +required
+	Generation int64 `json:"generation" protobuf:"bytes,2,name=generation"`
+
+	// ResourceSliceCount is the total number of ResourceSlices in the pool at this
+	// generation number. Must be higher than zero.
+	//
+	// Consumers can use this to check whether they have seen all ResourceSlices
+	// belonging to the same pool.
+	//
+	// +required
+	ResourceSliceCount int64 `json:"resourceSliceCount" protobuf:"bytes,3,name=resourceSliceCount"`
+}
+
+const ResourceSliceMaxSharedCapacity = 128
+const ResourceSliceMaxDevices = 128
+const PoolNameMaxLength = validation.DNS1123SubdomainMaxLength // Same as for a single node name.
+
+// Device represents one individual hardware instance that can be selected based
+// on its attributes.
+type Device struct {
+	// Name is unique identifier among all devices managed by
+	// the driver in the pool. It must be a DNS label.
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// Attributes defines the set of attributes for this device.
+	// The name of each attribute must be unique in that set.
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	Attributes map[QualifiedName]DeviceAttribute `json:"attributes,omitempty" protobuf:"bytes,2,rep,name=attributes"`
+
+	// Capacity defines the set of capacities for this device.
```

Comment:
https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1659327011 has become detached from code (yay).

The topics at hand were how partitionable devices fit and whether we have non-quantity capacity (or whether that should be a different field here).

Reply:
In https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1664530139 I outlined two ways of extending the API:
- A discriminator as in https://github.com/kubernetes-sigs/wg-device-management/pull/36#discussion_r1644884274 to warn old clients.
- A separate field.

I don't like the proposed discriminator because it's not clear what future values it should have as more features get added ("PartitionableDevicesOrSomethingElseAddedIn1.33"). Can we use a version number instead? See https://github.com/kubernetes/kubernetes/pull/125953.

When using a separate field, such a version number or discriminator are not needed. See https://github.com/kubernetes/kubernetes/pull/125952.

I previously said that I prefer the separate field because I didn't like the discriminator approach. But the version number is fine with me. That approach is simpler because we can extend the same `Device struct`. Existing Go code can work with that and doesn't need to be adapted unless it wants to support some new feature.

Therefore I prefer https://github.com/kubernetes/kubernetes/pull/125953. But to support it, we have to add `RequiredVersion` already in 1.31.

Note that neither of these two approaches are related to this `Capacity` here. That field is for something else.

Reply:
I replied on the thread here: https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1669060587, not sure if everyone will see it due to detachment, so pointing it out here.

I replied on the discriminator thing here: https://github.com/kubernetes/kubernetes/pull/125953/commits/7d809e1a33755dd52fcb55a26adc94bc29886d73#r1669026609

I am proposing a third option, but I think it's better...

Reply:
The discussion in https://github.com/kubernetes/kubernetes/pull/125953 points towards making `Device` a one-of ([Tim](https://github.com/kubernetes/kubernetes/pull/125953#discussion_r1669250444), [myself](https://github.com/kubernetes/kubernetes/pull/125953#discussion_r1669339106), [John](https://github.com/kubernetes/kubernetes/pull/125953#issuecomment-2215627948)).

I suggested `SimpleDevice` as first and only alternative in 1.31, but now like `BasicDevice` better: it's quite possible that future alternatives will inline it and thus use it as "base".

 I'll try that on for size by making that change here.

Reply:
As this comment thread is attached to "Capacities", let me also comment on that: the initial argument for splitting that out of "Attributes" was that we need "capacities" for future use cases like consumable capacity. That argument is now a bit weaker because we could add such a field on some new device struct when it is needed.

I'm still okay with already adding it now for the sake of consistency with other Kubernetes APIs and for consistency with future device types - just beware that the semantic is different. In the `BasicDevice` struct, capacities are just a special kind of attribute.

Reply:
This sounds OK to me, but just to make sure I understand -- the proposal is to still have:
```
type ResourceSlice struct {
    ...
    Devices []Device
}
```
But then make `Device` a one-of, e.g.:
```
type Device struct {
    Basic *BasicDevice
    Complex *ComplexDevice
}
```
I think I am OK with this -- it's basically pushing the one-of into the device type rather than on the slice as a whole (which is the one thing we actually care about discriminating on anyway).

And as you point out -- it then leaves the door open to change / extend the `Capacity` type in the future (if we want). 

So basically, define `BasicDevice` with what we know we need today, but leave the door open to define devices with a different structure in the future as we learn more / have more features we want to include.

Reply:
Exactly. I'm working on it now...

One change is that I am moving the "Name" up into `Device` because it's required. I didn't have it like that initially, but then some code which didn't need to know about device types had to "peek into" the `BasicDevice` to get the name.

Reply:
I push a PR update (squashed!) with that change. As far as I am concerned, this was the last API change that was required for 1.31.

I'm now going to implement the missing validation pieces.

Reply:
Awesome, yes, that sounds right, modulo the claim side api evolution discussion that is still open. I will have time to start another pass this afternoon.

Reply:
I finished the validation.go. The tests pass, but are not complete. Before I push, I'll run some more tests to see whether the API implementation is now complete.

While working on validation, I noticed a few minor gaps in the API documentation and naming.

Here's the diff, including the `Generation` update from Tim:
```patch
diff --git a/pkg/apis/resource/types.go b/pkg/apis/resource/types.go
index 2deca8e8d84..4a9bd6b5592 100644
--- a/pkg/apis/resource/types.go
+++ b/pkg/apis/resource/types.go
@@ -87,12 +87,13 @@ type ResourceSliceSpec struct {
 	// objects with a certain driver name.
 	//
 	// Must be a DNS subdomain and should end with a DNS domain owned by the
-	// vendor of the driver.
+	// vendor of the driver. This field is immutable.
 	//
 	// +required
 	Driver string
 
 	// Pool describes the pool that this ResourceSlice belongs to.
+	// This field is immutable.
 	//
 	// +required
 	Pool ResourcePool
@@ -107,6 +108,7 @@ type ResourceSliceSpec struct {
 	// resources available.
 	//
 	// Exactly one of NodeName, NodeSelector and AllNodes must be set.
+	// This field is immutable.
 	//
 	// +optional
 	// +oneOf=NodeSelection
@@ -148,14 +150,18 @@ type ResourcePool struct {
 	// +required
 	Name string
 
-	// Generation must get increased in all ResourceSlices of a pool whenever resource
-	// definitions change. A consumer must only use definitions from ResourceSlices
-	// with the highest generation number and ignore all others.
+	// Generation tracks the change in a pool over time. Whenever a driver
+	// changes something about one or more of the resources in a pool, it
+	// must change the generation in all ResourceSlices which are part of
+	// that pool. Consumers of ResourceSlices should only consider
+	// resources from the pool with the highest generation number. The
+	// generation may be reset by drivers, which should be fine for
+	// consumers, assuming that all ResourceSlices in a pool are updated to
+	// match or deleted.
 	//
-	// This mechanism enables consumers to detect pools that are in an incomplete state.
-	// It is not sufficient to determine whether the pool is up-to-date. Because that
-	// race is unavoidable, drivers may start publishing a pool with zero as initial
-	// generation number when there are no slices for that pool with a higher number.
+	// Combined with ResourceSliceCount, this mechanism enables consumers to
+	// detect pools which are comprised of multiple ResourceSlices and are
+	// in an incomplete state.
 	//
 	// +required
 	Generation int64
@@ -353,6 +359,12 @@ type DeviceClaim struct {
 	Config []DeviceClaimConfiguration
 }
 
+const (
+	DeviceRequestsMaxSize    = AllocationResultsMaxSize
+	DeviceConstraintsMaxSize = 32
+	DeviceConfigMaxSize      = 32
+)
+
 // DeviceRequest is a request for devices required for a claim.
 // This is typically a request for a single resource like a device, but can
 // also ask for several identical devices.
@@ -432,6 +444,10 @@ type DeviceRequest struct {
 	AdminAccess bool
 }
 
+const (
+	DeviceSelectorsMaxSize = 32
+)
+
 type DeviceCountMode string
 
 // Valid [DeviceRequest.CountMode] values.
@@ -917,7 +933,7 @@ type DeviceClassSpec struct {
 	//
 	// +optional
 	// +listType=atomic
-	Config []ClassConfiguration
+	Config []DeviceClassConfiguration
 
 	// Only nodes matching the selector will be considered by the scheduler
 	// when trying to find a Node that fits a Pod when that Pod uses
@@ -935,8 +951,8 @@ type DeviceClassSpec struct {
 	SuitableNodes *core.NodeSelector
 }
 
-// ClassConfiguration is used in DeviceClass.
-type ClassConfiguration struct {
+// DeviceClassConfiguration is used in DeviceClass.
+type DeviceClassConfiguration struct {
 	DeviceConfiguration // inline
 }
 
```

Reply:
Pushed. Next TODO:
- split up the "adapt implementation to new API" PR by component
- investigate a strange pkg/api/testing TestDefaulting failure
- DRAControlPlaneController feature gate